### PR TITLE
Add CommandContext into all tuples.

### PR DESCRIPTION
**Pull request by dpoltavets**
Closes: #2199, #2239.

---
This change is [Reviewable](https://reviewable.io/reviews/telstra/open-kilda/2253)

### CODE REVIEW

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/AbstractTopology.java, line 35 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-LceO1U23BVcjh5bhHKU:-Lcj8N1-5BEQvx8I1hTk:b-rfkvmf) ([raw file](https://github.com/telstra/open-kilda/blob/a6ac3343a29f8934d2f872d8d50f9cc45584e273/services/wfm/src/main/java/org/openkilda/wfm/topology/AbstractTopology.java#L35)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Can we drop `KeyValueKafkaRecordTranslator`?
</blockquote></details>

No, we can't. `KafkaRecordTranslator` is used when we create spout for OpenTsdbTopology. In this topology we receive InfoData instead of Message, so we can't use `new` spout.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/AbstractTopology.java, line 215 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-LcfTVQX2JHnfQKLcRiZ:-LcjAijwAHWi8AMjoEMU:bss4a2b) ([raw file](https://github.com/telstra/open-kilda/blob/a6ac3343a29f8934d2f872d8d50f9cc45584e273/services/wfm/src/main/java/org/openkilda/wfm/topology/AbstractTopology.java#L215)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Why for you keep this "deprecated" spout builder?
</blockquote></details>

Removed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/AbstractTopology.java, line 226 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-LceORZV8IicGpr-jJY3:-LcjAlNl5qYLcxDhDnDk:b-896fix) ([raw file](https://github.com/telstra/open-kilda/blob/c16e919fef06b25ed5d542725e9624b0a4d776e8/services/wfm/src/main/java/org/openkilda/wfm/topology/AbstractTopology.java#L226)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Wrap topic into list and call `buildKafkaSpout(List<String> topics, String spoutId)`.
</blockquote></details>

Done.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/nbworker/bolts/FlowOperationsBolt.java, line 135 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-LcfVihq3ONd95TvuKkl:-LcjAptU3MdoaIvkvxvx:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/c16e919fef06b25ed5d542725e9624b0a4d776e8/services/wfm/src/main/java/org/openkilda/wfm/topology/nbworker/bolts/FlowOperationsBolt.java#L135)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Do we really need correlation-id here? The only "catch" clause, that catch this exception, use "type", "message", and "description" fields from this exception object.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/nbworker/bolts/PersistenceOperationsBolt.java, line 53 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-LcfYFylD9BwhuGggLja:-LcjAsXi4fhMROeXDgDS:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/c16e919fef06b25ed5d542725e9624b0a4d776e8/services/wfm/src/main/java/org/openkilda/wfm/topology/nbworker/bolts/PersistenceOperationsBolt.java#L53)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Drop it and replace all caller with `getCurrentTuple()`.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/nbworker/bolts/ResponseSplitterBolt.java, line 38 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-LcfYgyl5Ddnvf-n7Myq:-LcjAvy0-XbC7w-P8VuO:b-896fix) ([raw file](https://github.com/telstra/open-kilda/blob/c16e919fef06b25ed5d542725e9624b0a4d776e8/services/wfm/src/main/java/org/openkilda/wfm/topology/nbworker/bolts/ResponseSplitterBolt.java#L38)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Should be dropped now.
</blockquote></details>

Done.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/nbworker/bolts/RouterBolt.java, line 75 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-LcfZF3V9Y4pa_shok3K:-LcjAzNf2Qjnm2fS0lu5:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/c16e919fef06b25ed5d542725e9624b0a4d776e8/services/wfm/src/main/java/org/openkilda/wfm/topology/nbworker/bolts/RouterBolt.java#L75)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

`processRequest`(all of them)  can call `getCommandContext()` by himself.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/CacheBolt.java, line 134 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-Lcf_KgP3tJeiABPDjYc:-LcjB0uy4lGGJdF0kC4f:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/c16e919fef06b25ed5d542725e9624b0a4d776e8/services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/CacheBolt.java#L134)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

why not `getCommandContext()`?
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/SpeakerBolt.java, line 76 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-LcfbNbWBCfeccG7j9Pz:-LcjB3UNE-4nYU8JFkh5:b-896fix) ([raw file](https://github.com/telstra/open-kilda/blob/c16e919fef06b25ed5d542725e9624b0a4d776e8/services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/SpeakerBolt.java#L76)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Remove wrapping in `Values` into log calls.
</blockquote></details>

Done.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/SpeakerBolt.java, line 93 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-LcfbZi89VniAlecUcvm:-LcjB62CEJYG-bgKeTwZ:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/a6ac3343a29f8934d2f872d8d50f9cc45584e273/services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/SpeakerBolt.java#L93)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

`try { ... } finally { outputCollector.ack(); }` must stay.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/SpeakerBolt.java, line 125 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-Lcfbx7W96vv2jqvTt3M:-LcjB83X6ZCr6qjJBoYK:bcgtdcu) ([raw file](https://github.com/telstra/open-kilda/blob/c16e919fef06b25ed5d542725e9624b0a4d776e8/services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/SpeakerBolt.java#L125)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

O_o... can it be done in simplier way?
</blockquote></details>

Yes, it can be. Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/AbstractTopology.java, line 35 at r1](https://reviewable.io/reviews/telstra/open-kilda/2253#-LceO1U23BVcjh5bhHKU:-Ld7rDe078e5eTGJd2-M:b-22s2aj) ([raw file](https://github.com/telstra/open-kilda/blob/a6ac3343a29f8934d2f872d8d50f9cc45584e273/services/wfm/src/main/java/org/openkilda/wfm/topology/AbstractTopology.java#L35)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

This means we need one more value decoder i.e. we should be able to define value decoder on topology level, during spout configuration. And drop `KafkaRecordTranslator`. Actually we already can pass value translator. So we need only one more `translator`.
</blockquote></details>

I've created new `translator` for InfoData and removed `KafkaRecordTranslator`.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/event/PortEventRouterBolt.java, line 46 at r3](https://reviewable.io/reviews/telstra/open-kilda/2253#-Ld4DxoU3AcsqtvOY22I:-Ld7rV-R5QWgMfcvcKxS:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/17ac85e8b17e042808a7a5a67f6f75d1e05772c4/services/wfm/src/main/java/org/openkilda/wfm/topology/event/PortEventRouterBolt.java#L46)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

use `pullValue(...)` if you have `AbstractBolt` in ancestors.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/flow/FlowTopology.java, line 75 at r3](https://reviewable.io/reviews/telstra/open-kilda/2253#-Ld4FK-Q-mPkPRCDkMWV:-Ld7rXz76q0tmRCq8eBF:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/17ac85e8b17e042808a7a5a67f6f75d1e05772c4/services/wfm/src/main/java/org/openkilda/wfm/topology/flow/FlowTopology.java#L75)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

`SPEAKER_KAFKA_SPOUT` here and `FLOW_KAFKA_SPOUT` on the next line... E???
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/kafka/InfoDataDeserializer.java, line 33 at r4](https://reviewable.io/reviews/telstra/open-kilda/2253#-Ld87OB76KcI_5lZEG6d:-LdDf4sz9Fn4PxrDXwzF:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/113837581f227af6509c7fff8b16ab21e0e57cb6/services/wfm/src/main/java/org/openkilda/wfm/kafka/InfoDataDeserializer.java#L33)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

copy & paste in programming is evil. You can make abstract class(from this class) with astract method `protected T decode(byte[] data);` and implement 2 different classed that implement this method for `Message` and for `InfoData` types.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/opentsdb/OpenTsdbTopology.java, line 91 at r4](https://reviewable.io/reviews/telstra/open-kilda/2253#-Ld8800F7kjdp4ZLqM0u:-LdDf7iS5sRC1uYyX9Rr:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/628ea55d78405a77fdfe80e3ad259525d2376e1e/services/wfm/src/main/java/org/openkilda/wfm/topology/opentsdb/OpenTsdbTopology.java#L91)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

overkill... base class should know nothing about child class logic and responsibilities.

Use our common builder (`getKafkaSpoutConfigBuilder`) and redefine required fields. 
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/opentsdb/bolts/DatapointParseBolt.java, line 55 at r4](https://reviewable.io/reviews/telstra/open-kilda/2253#-Ld89figEOhvFJLAODPl:-LdDf9aQ9pO-M7sbdfw2:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/113837581f227af6509c7fff8b16ab21e0e57cb6/services/wfm/src/main/java/org/openkilda/wfm/topology/opentsdb/bolts/DatapointParseBolt.java#L55)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

"else" branch is required here. We should be able to know if some data goes into incorrect place.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/kafka/InfoDataDeserializer.java, line 33 at r4](https://reviewable.io/reviews/telstra/open-kilda/2253#-Ld87OB76KcI_5lZEG6d:-LdEJSCrB-ztRbLt9Vlv:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/113837581f227af6509c7fff8b16ab21e0e57cb6/services/wfm/src/main/java/org/openkilda/wfm/kafka/InfoDataDeserializer.java#L33)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

...no

move exception handling into `Deserializer`. Here you should have only:
```
protected InfoData jsonDecode(byte[] data) {
    return Utils.MAPPER.readValue(data, InfoData.class);
}
```
Plus correct "raises".

Everything else is the responsibility of `Deserializer<T>`.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/AbstractTopology.java, line 265 at r5](https://reviewable.io/reviews/telstra/open-kilda/2253#-LdE07JNEWqCmgxyJS_p:-LdEJXXs7edxKyYfq8Dn:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/e996698ff23479eadc4dc4d476ae489f2f453b60/services/wfm/src/main/java/org/openkilda/wfm/topology/AbstractTopology.java#L265)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Don't do it... one more time. Do it on the caller side.

```
getKafkaSpoutConfigBuilder(Collections.singletonList(topicName), "spoitABC)
    .setFirstPollOffsetStrategy(KafkaSpoutConfig.FirstPollOffsetStrategy.UNCOMMITTED_EARLIEST)
    .setRecordTranslator(new InfoDataTranslator());
```

It there any issue with redefining already defined fields in the builder?
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/event/OfEventWfmTopology.java, line 122 at r9](https://reviewable.io/reviews/telstra/open-kilda/2253#-LfdX54sEnYIhMc6Yj3e:-LfrYGeK1aZXqfwIi1Wi:bss4a2b) ([raw file](https://github.com/telstra/open-kilda/blob/c0ceabd50d5ee6ee1999327d6a930ecb6b4115d4/services/wfm/src/main/java/org/openkilda/wfm/topology/event/OfEventWfmTopology.java#L122)):*
<details><summary><i>Previously, IlyaMoiseev (Ilya Moiseev) wrote…</i></summary><blockquote>

> ```
> createCtrlBranch(builder, ctrlTargets);
> ```
why also not remove corresponds `ctrlTargets` stuff?
</blockquote></details>

Removed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/flow/FlowTopology.java, line 188 at r9](https://reviewable.io/reviews/telstra/open-kilda/2253#-LfdZoLD5AG-UcZbqdVQ:-LfrYMMy9EmLdxnCldiw:bss4a2b) ([raw file](https://github.com/telstra/open-kilda/blob/a6352b81d6007c26322dc183af368a78087e4414/services/wfm/src/main/java/org/openkilda/wfm/topology/flow/FlowTopology.java#L188)):*
<details><summary><i>Previously, IlyaMoiseev (Ilya Moiseev) wrote…</i></summary><blockquote>

> ```
> ctrlTargets
> ```
the same
</blockquote></details>

Removed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/flow/bolts/SplitterBolt.java, line 85 at r9](https://reviewable.io/reviews/telstra/open-kilda/2253#-Lfd_ACb3O_-7VfDKQ3I:-LfrYPZm1oBXw6KEYaLW:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/c0ceabd50d5ee6ee1999327d6a930ecb6b4115d4/services/wfm/src/main/java/org/openkilda/wfm/topology/flow/bolts/SplitterBolt.java#L85)):*
<details><summary><i>Previously, IlyaMoiseev (Ilya Moiseev) wrote…</i></summary><blockquote>

It's safer to move this into try block
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/SpeakerBolt.java, line 66 at r9](https://reviewable.io/reviews/telstra/open-kilda/2253#-LfdaJun4SPvi4c9VGG1:-LfrYRea6WKYWVJGgQbw:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/c0ceabd50d5ee6ee1999327d6a930ecb6b4115d4/services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/SpeakerBolt.java#L66)):*
<details><summary><i>Previously, IlyaMoiseev (Ilya Moiseev) wrote…</i></summary><blockquote>

Safer to move this into try section with error logging  catch block
</blockquote></details>

Fixed.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/flow/FlowTopology.java [hunk 1]
```diff
@@ -66,25 +60,22 @@ public StormTopology createTopology() throws NameCollisionException {
         logger.info("Creating FlowTopology - {}", topologyName);
 
         TopologyBuilder builder = new TopologyBuilder();
-        final List<CtrlBoltRef> ctrlTargets = new ArrayList<>();
         Integer parallelism = topologyConfig.getParallelism();
 
         /*
          * Spout receives all Northbound requests.
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/flow/FlowTopology.java]: Shouldn't the comment be updated along with the renaming NORTHBOUND_KAFKA_SPOUT => FLOW_KAFKA_SPOUT?

File: services/wfm/src/main/java/org/openkilda/wfm/AbstractBolt.java [hunk 2]
```diff
@@ -134,7 +134,7 @@ protected CommandContext setupCommandContext() {
         } catch (PipelineException e) {
             context = new CommandContext().fork("trace-fail");
 
-            log.debug("The command context is missing in input tuple received by {} on stream {}:{}, execution context"
+            log.warn("The command context is missing in input tuple received by {} on stream {}:{}, execution context"
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply services/wfm/src/main/java/org/openkilda/wfm/AbstractBolt.java]: Added the exception details.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/flow/FlowTopology.java (hunk 1 above)
Comment (dpoltavets) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/flow/FlowTopology.java]: Renamed.

**Pull request merged**